rfcs/0002-tracing-cache.md

```diff
@@ -0,0 +1,95 @@
+
+Ideally our binary format would have the following properties:
+* Ability to skip records which are filtered out without parsing
+* Zero copy to allow for a subset of a record's fields to be read when deciding whether to filter it. (Alternatively, easily separated hot/cold storage)
+* Little to no overhead for the schema, as there are no optional fields and backwards compatibility can be amortized into the collection
+* Processable as a linear stream with minimal cache misses, ideally not requiring the entire file to be loaded at once.
+* Move redundant data outside of the Trace (eg: Block Number)
+
+JSON is the format the data is delivered in today. JSON is neither compressed nor fast: https://eng.uber.com/trip-data-squeeze/. JSON does not have any of the properties of our ideal format. The entirety of a JSON file needs to be parsed whether or not all of it is actually used. JSON's self-description and flexibility come at a large cost for data which has a schema. As much of the data in a trace is binary data or data which does not have an established representation in JSON, additional overheads are incurred in encoding representations of traces in the primitives that JSON supports.
+
+Flatbuffers seems like a good initial candidate here which makes reasonable tradeoffs between skip/zero-copy and compression ratios. Significant gains over Flatbuffers are possible with a custom format, but this is not likely a good place to spend time in the short term. 
+
+This is subject to change if it is decided that filtering/indexing should happen within postgres.
+
+## Compatibility
+
+This change is backwards compatible. Existing code can continue to use the parity tracing API. Because the cache is local, each indexing node may delete the cache should the format or implementation of caching change. In this case of invalidated cache the code will fall back to existing methods for retrieving a trace and repopulating the cache.
+
+## Drawbacks and Risks
+
+In some cases, a graph node may not frequently get a cache hit when requesting a trace. If cache hits are relatively rare, the additional cost of pre-processing the data and caching it would only add to the overhead that exists today. It is currently up to a combination of the implementation details of block_stream and the set of subgraphs being indexed as to whether this strategy would be a win for a particular graph node. It may be necessary to monitor statistics about the behavior of the caching layer to keep tabs on whether this change continues to be a win in a production environment.
+
+To mitigate this risk, it may be worth frontloading work on the cache hierarchy to gather statistics about block trace access patterns and how frequently cache hits are expected for existing graph nodes.
+
+## Alternatives
+
+A general purpose etherium cache has been suggested. It is likely that much of the work here would be able to be repurposed for such a cache.
```

Comment:
We could restructure the existing `EthereumCallCache` to be a more general `EthereumCache` that also covers traces. This is a decision we can make when writing the engineering plan for the feature though; whether we do it or not shouldn't matter at all in terms of effort, since the surface area of both caches is narrow.